Tasks/SonarQubePostTest/SonarQubeBuildBreaker.ps1

```diff
@@ -0,0 +1,176 @@
+    }
+}
+
+#
+# Reads the task id from task-report.txt that is dropped by the sonar-scanner to give the task id
+#
+function FetchTaskIdFromReportFile
+{
+    param ([string]$reportTaskFile)
+
+    $content = [System.IO.File]::ReadAllText($reportTaskFile);
+    $matchResult = [System.Text.RegularExpressions.Regex]::Match($content, "ceTaskId=(.+)");    
+
+    if (!$matchResult.Success -or !$matchResult.Groups -or $matchResult.Groups.Count -ne 2)
+    {
+        throw "Could not find the task Id in $reportTaskFile. If you are using SonarQube 5.3 or above, please raise a bug."
+    }
+
+    $taskId = $matchResult.Groups[1].Value
+    Write-Verbose "The analysis is associated with the task id $taskId"
+
+    return $taskId
+}
+
+#
+# Queries the server to determine if the task has finished, i.e. if the quality gate has been evaluated
+#
+function QueryAnalysisFinished
+{
+    param ([string]$taskId)
+    
+    # response is in json and ps deserialize it automatically
+    $response = InvokeGetRestMethod "/api/ce/task?id=$taskId" $true    
+    $status = $response.task.status
+    
+    Write-Verbose "The task status is $status"
+
+    if (!$status)
+    {
+        throw "Could not determine the task status - please raise a bug."
+    }
+    
+    return $status -eq "success"   
+}
+
+#
+# Query the server to determine the analysis id associated with the current analysis
+#
+function QueryAnalysisId
+{
+    param ([string]$taskId)
+       
+    $response = InvokeGetRestMethod "/api/ce/task?id=$taskId" $true    
+    return $response.task.analysisId      
+}
+
+#
+# Returns the path to the report-task.txt file containing task details
+#
+function GetTaskStatusFile
+{    
+    $sonarDir = GetSonarScannerDirectory 
+    $reportTaskFile = [System.IO.Path]::Combine($sonarDir, "report-task.txt");
+    
+    if (![System.IO.File]::Exists($reportTaskFile))
+    {
+        Write-Verbose "Could not find the task details file at $reportTaskFile"
+        throw "Cannot determine if the analysis has finished in order to break the build. Possible causes: your SonarQube server version is lower than 5.3 - for more details on how to break the build in this case see http://go.microsoft.com/fwlink/?LinkId=722407"
+    }
+
+    return $reportTaskFile
+}
+
+#
+# Pools the server until current analysis is complete or a timeout is hit. Returns the analysis id.
+#
+function WaitForAnalysisToFinish
+{    
+    Write-Host "Waiting on the SonarQube server to finish processing in order to determine the quality gate status."
+       
+    $reportPath = GetTaskStatusFile    
+    $taskId = FetchTaskIdFromReportFile $reportPath
+        
+    $command = { QueryAnalysisFinished $taskId }
+
+    $numRetries = GetNumberOfPollingRetries
+    $taskFinished = RetryUntilTrue $command -maxRetries $numRetries -retryDelay 1 
+
+    if (!$taskFinished)
+    {
+        throw "The analysis did not complete after checking the task status $numRetries times."
```

Comment:
I'm not sure the number of retries is relevant at this point.
"Timed out waiting for the analysis to complete on the SonarQube server." might be better - it makes it clear in the error message where the processing is happening.